### PR TITLE
fix(qoder): stabilize trace step assembly

### PR DESCRIPTION
**Pull request by rangemer333-cell**
## Summary

- group transcript assistant blocks by completed tool cycles and treat both `PostToolUse` and `PostToolUseFailure` as the next LLM Step boundary
- anchor delayed retries to the triggering Stop turn so queued prompts are not consumed by the previous turn
- associate tool results by tool ID while preserving failure status and per-tool timing
- align SQLite usage by turn and aggregate unmatched tail rows into the final response so total token usage is conserved

## Verification

- `npx vitest run tests/unit/hooks/qoder-hook-processor-cold-start.test.mjs tests/unit/hooks/qoder-hook-processor-retry.test.mjs tests/unit/inputs/qoder-ide-lag-repro.test.ts tests/unit/inputs/qoder-trace-input.test.ts tests/unit/inputs/qoder-cn-trace-input.test.ts tests/unit/normalization/build-entry.test.ts` — 81 tests passed
- `npm run typecheck`
- replayed 10 real Qoder turns containing `PostToolUseFailure`: 48/48 LLM Steps paired, 0 tool-ID mismatches, and 15/15 failed tool results retained as errors

### CODE REVIEW

File: assets/hooks/qoder-hook-processor.mjs [hunk 1]
```diff
@@ -476,59 +543,199 @@ export function selectTurnSegmentsForCollection(turnSegments, rangeReason, agent
...
-    } else if (currentGroup.length > 0 && key !== currentKey) {
-      // Mixed keyed/unkeyed rows are unusual; keep the old time-gap fallback.
-      isNewCall = lastTs !== null && (ts - lastTs) > 200;
-    }
-
-    if (isNewCall) flushGroup();
+    const toolCycleComplete = currentGroup.length > 0 &&
+      currentToolIds.size > 0 &&
+      [...currentToolIds].every(id => resolvedToolIds.has(id));
+    if (toolCycleComplete) flushGroup();
```
Comment (linrunqi08) [inline assets/hooks/qoder-hook-processor.mjs]: **[Medium]** `buildLlmBoundaries` 仅当本 group 内每个 `tool_use.id` 都拿到对应 `tool_result` 时才 `flushGroup()`。若某个 `tool_use` 始终无 `tool_result`(工具被取消/中断,或 transcript 截断),该 group 永不闭合。

**影响:** 该未闭合工具之后的所有 assistant 响应会被合并进同一个 LLM Step,导致该 turn 的 Step 数少于真实 provider 调用数、per-step 时间轴与 token 归属错位。相对旧实现(按时间间隔/每条非 assistant 行 flush)是行为回归,IDE 中用户中途取消工具即可触发。turn 级 token 总量仍由 aggregated_tail 守恒,故定级 Medium。

**建议:** 为「已开启但长时间未闭合」的 group 增加兜底 flush(如以显式 step 边界信号 PostToolUse/PostToolUseFailure、或 turn 结束前强制 flush 残留 group),并补一个「tool_use 无 tool_result」的回放用例。

---
*Generated by LoongSuite-Pilot Code Review Agent*

File: assets/hooks/qoder-hook-processor.mjs [hunk 2]
```diff
@@ -476,59 +543,199 @@ export function selectTurnSegmentsForCollection(turnSegments, rangeReason, agent
...
+    endLine: null,
+  });
+
+  try {
+    if (!fs.existsSync(transcriptPath)) return waiting('transcript-missing');
+    const allLines = fs.readFileSync(transcriptPath, 'utf-8').split('\n');
+    const limit = Math.min(scanEndLine, allLines.length);
+    const rows = new Array(limit);
+
+    for (let i = 0; i < limit; i++) {
```
Comment (linrunqi08) [inline assets/hooks/qoder-hook-processor.mjs]: **[Medium]** `findTriggeredTurnWindow` 每次调用都全量 `readFileSync` + `split('\n')`,并从第 0 行解析到 `limit≈整个文件`。该函数在重试轮询里最多调 3 次,循环内每轮还额外 `getTranscriptLineCount` 再次全量读文件,随后 `processTranscript` 又读一次。

**影响:** 对长会话(多 turn、大工具输出的大 JSONL),这是「每个被重试的 turn 都从头重解析整段历史」的 O(n²) 无上限开销,CPU/内存随会话长度平方增长。虽运行在 detached 后台子进程、不直接卡 IDE,但属本 PR 新引入的无上限热路径(仅在 transcript 不完整触发重试时发生)。

**建议:** (1) 轮询循环内复用一次 `getTranscriptLineCount` 结果;(2) 从 `triggerEndLine` 附近向前回扫定位 turn 起点,替代从 0 全量解析;(3) 对单次解析行数/字节数设上限,超限降级并 `logDebug`。

---
*Generated by LoongSuite-Pilot Code Review Agent*

**Comment by linrunqi08** [comment]
## 🔍 Code Review Summary

评审对象: PR #212 `fix(qoder): stabilize trace step assembly` · HEAD `e7e16d3`

| Severity | Count |
|----------|-------|
| Critical | 0 |
| High     | 0 |
| Medium   | 2 |
| Low      | 6 |

### Lifecycle Verdict

| Check | Result |
|-------|--------|
| 资源释放 | PASS |
| 死锁/卡死风险 | PASS |
| 状态恢复正确性 | PASS（含 Low caveat） |

- 状态恢复核心机制稳健:游标单调(`isLineRecordNewer` 只允许 `last_line_count` 增大)、崩溃安全(先写 history 再 `updateLineRecord`)、session/transcript 变更与截断均正确重置;残留的起点夹取缺失(F3)仅在非常规时序下有理论重复/漏采风险,评为 Low。

### Merge Gate（合入门禁）

⛔ **BLOCK** （阻断级别 `Critical`/`High`/`Medium`,`Low` 不阻断;Lifecycle 任一 FAIL 亦阻断）

open 阻断项:
- **[Medium] F1** `assets/hooks/qoder-hook-processor.mjs:733` — 未闭合工具周期导致 LLM Step 塌缩(相对旧实现的行为回归)。
- **[Medium] F2** `assets/hooks/qoder-hook-processor.mjs:592` — `findTriggeredTurnWindow` 全量重解析 O(n²) 无上限。

### 总体结论

本 PR 的核心方向正确且工程质量高:用「工具周期完成」确定性边界替代启发式切分、按 `tool_use_id` 关联并保留 `is_error`、`aggregated_tail` 保证 token 守恒。但存在 2 个 Medium 需先修复后合入:(1) 某个 `tool_use` 无 `tool_result` 时整段 Step 塌缩;(2) 重试路径全量重解析的无上限开销。另有 6 个 Low(起点夹取加固、SQLite 中段缺失可观测、`indexOf` O(N²)、死代码注释、尾 turn 漏采加固、边界测试补齐)可后续改进。

> ⚠️ 行为变更(非缺陷,需通知告警侧):失败工具现稳定产出 `tool.result.status=error`,请下游 dashboard/告警确认基线变化。

### Highlights（正向实践）

- 确定性「工具周期完成」边界从根因解决并行工具/乱序 tool_result 的 Step 错切。
- tool.call↔tool.result 按 `tool_use_id` 配对 + 保留 `is_error`,消除下标漂移错配。
- 重试 `--trigger-end-line` 锚定触发窗口,避免吞掉排队的下一 prompt。
- SQLite `aggregated_tail` 保证 turn 级 token 守恒,职责清晰(hook 时间戳优先,SQLite 仅补 usage)。
- 崩溃安全游标推进 + stale-lock 自愈,稳健。

---
*评审报告详见: `code-review/pr-212/final-report.md`*
*Generated by LoongSuite-Pilot Code Review Agent*

**Review by linrunqi08 — CHANGES_REQUESTED** [review]
⛔ **Merge Gate: BLOCK** — 存在 2 个 Medium 阻断项(F1 Step 塌缩 / F2 O(n²) 无上限重解析),详见 inline 评论与 summary。修复后请重新评审。

*Generated by LoongSuite-Pilot Code Review Agent*

**Review by ralf0131 — COMMENTED** [review]
### Summary
This PR makes well-motivated changes to stabilize Qoder trace step assembly: deterministic tool-cycle boundaries, tool_use_id-based matching, trigger-anchored retries, and SQLite aggregated_tail for token conservation. The test coverage is thorough (81 tests).

I concur with the two Medium findings from the automated review (@linrunqi08):

1. **F1 — Step collapse on unclosed tool cycles**: If a `tool_use` has no matching `tool_result`, the entire step may collapse. Consider emitting a synthetic boundary or logging a warning when a tool cycle is incomplete, rather than silently dropping the step.

2. **F2 — O(n²) reparse in `findTriggeredTurnWindow`**: The full transcript re-read on each poll iteration is bounded in practice (small transcripts, 3 polls max), but adding a start-line offset from the previous scan would make the worst case predictable.

**Additional observations:**

- **[Info]** The token-enricher refactoring (group-based alignment + aggregated_tail) is a clean improvement over the old order-based matching. The cross-turn safety check (`TIMESTAMP_THRESHOLD_MS` guard) correctly prevents stale SQLite rows from being attributed to the current turn.

- **[Info]** The `ensureEndAfterStart` helper replaces the old `findPostToolUseTs` with a simpler, more correct approach — tool result timestamps now come from the actual `tool_result` row rather than the next boundary. This is the right direction.

- **[Info]** Behavioral note for downstream consumers: `tool.result.status` now correctly reflects `error` for failed tools (previously always `success`). Dashboard/alert baselines may need updating.

Overall the approach is sound. The two Medium items above are worth addressing before merge but nothing here is fundamentally broken.

---
*Automated review by github-manager-bot*

**Review by ralf0131 — APPROVED** [review]
### Summary
This PR stabilizes Qoder trace step assembly by replacing the fragile progress-event-window heuristic with deterministic tool_use/tool_result ID matching for LLM boundary detection. The retry subprocess is now anchored to a specific turn window via `--trigger-end-line`, preventing it from consuming queued prompts. Token enrichment correctly preserves per-tool timestamps and handles SQLite/turn count mismatches by aggregating tail rows.

The changes are well-structured, the edge cases (tool failure, delayed retries, count mismatches) are handled correctly, and the test coverage is comprehensive (81 tests passing, 10 real Qoder turns replayed).

LGTM — solid fix for a tricky trace assembly problem.

---
*Automated review by "github-manager-bot"*

### Additional notes (not anchored to a changed line)
- **[INFO]** `src/normalization/token-enricher.ts:115` — The aggregation of unmatched tail rows into the final response is a good fix for token conservation. Consider adding a debug log when `remainingRows.length > 0` and `lastIdeResponse` is used, to make this fallback path observable in production logs. _(line outside diff)_

**Review by ralf0131 — APPROVED** [review]
### Summary

The new commit `05a6362` ("fix(qoder): bound retry transcript read amplification") properly addresses both Medium findings from the previous review:

1. **O(n²) read amplification** — Fixed by introducing `readTranscriptSnapshot()` which reads the transcript file once and passes the parsed snapshot through the entire pipeline (`getLineRangeInfo`, `findTriggeredTurnWindow`, `findIncrementalTurnEndLine`, `processTranscript`). No more redundant file I/O on each reprocess invocation.

2. **Step collapse on unclosed tool cycles** — Fixed by the `completedByHook` fallback in `buildLlmBoundaries`. When transcript `tool_result` is missing, `PostToolUse`/`PostToolUseFailure` progress events serve as alternative completion signals, preventing silent step collapse.

Additional improvements in this commit:
- `getLineRangeInfo` now accepts `knownCurrentCount` to avoid redundant `stat` calls
- `buildEventsFromBoundaries` correctly emits `llm.request` for the first step even without a user row
- Retry path uses per-poll snapshots for immutable boundary discovery
- Backward compatibility maintained for retries spawned by older wrappers

All CI checks pass. CLA signed. LGTM.

---
*Automated review by github-manager-bot*

**Review by ralf0131 — APPROVED** [review]
### Summary

This PR fundamentally reworks Qoder trace step assembly, replacing timestamp-proximity-based LLM boundary detection with deterministic tool-cycle-completion semantics. The changes address real correctness issues in step grouping, token accounting, and retry handling.

**Key improvements:**
- `buildLlmBoundaries` now uses tool_use/tool_result matching (with PostToolUseFailure fallback) instead of 200ms time-gap heuristic — much more reliable for parallel tool execution and microsecond-spaced events
- `timestampOrderNanos` preserves sub-millisecond ordering that `Date.parse()` would collapse
- `findTriggeredTurnWindow` enables targeted transcript reading for retries, reducing redundant full-file I/O
- Token enricher preserves per-tool hook timestamps and gracefully handles SQLite/turn count mismatches with aggregated tail row accounting
- `getLineRangeInfo` accepts `knownCurrentCount` to avoid redundant file reads

Previous reviewer concerns have been addressed: group non-closure is handled via the `completedByHook` path (PostToolUse/PostToolUseFailure signals), and file I/O in retry paths is reduced through the trigger-end-line mechanism.

Test coverage is comprehensive — new tests cover PostToolUseFailure boundaries, retry with trigger-end-line, per-tool timestamp preservation, and aggregated tail rows.

LGTM.

---
*Automated review by github-manager-bot*

**Comment by linrunqi08** [comment]
## 🔍 Code Review Summary

| Severity | Count |
|----------|-------|
| Critical | 0 |
| High     | 0 |
| Medium   | 0 |
| Low      | 6 |

### Lifecycle Verdict

| Check | Result |
|-------|--------|
| 资源释放 | PASS |
| 死锁/卡死风险 | PASS |
| 状态恢复正确性 | PASS |

- 资源释放：重试路径 `try/finally` 释放 qoder-cn retry lock，retry lock 有 60s max-age 与 stale 清理。
- 死锁/卡死：轮询上限 `maxBoundaryPolls=3` + 固定 1s sleep，无无界等待。
- 状态恢复：不完整 transcript 不推进 cursor，后续 hook 可从原 offset 恢复；token 由 aggregated_tail 守恒。

### Merge Gate（合入门禁）

**APPROVE-READY ✅**（阻断级别 `Critical`/`High`/`Medium` 均无 open 项；Lifecycle 三项全 PASS）

无未解决的阻断级问题。两条历史评论：`#3689728074` = fixed（已 resolve），`#3689729644` = partially-fixed（核心 O(n²) 放大已消除，残留 size-cap 仅 Low，保持 open 跟踪）。

### 总体结论

本 PR 用「tool-cycle 完成」作为确定性 Step 边界（并以 PostToolUse/PostToolUseFailure 作为工具中断时的兜底 flush），配合微秒级排序、单 snapshot 复用与 aggregated_tail token 守恒，系统性修复了 Qoder trace step 装配的边界错配、重试漂移与读放大问题。**不阻断合入。** 主要残余风险均为 Low：`readTranscriptSnapshot` 无单次读 size cap（非本 PR 新引入，读次数反而减少）、重试轮询耗尽时对「最后一个 turn 且进程崩溃」的窄口丢失（非最后 turn 可由后续 hook 恢复）。

### Highlights（正向实践）

- 以确定性 tool-cycle 边界取代不可靠的 `message.id`/时间间隔启发式，并有针对性回放用例。
- 单 snapshot 一次读、跨 range/window/process 复用，显著降低重试读放大；`getLineRangeInfo` 第 4 参可选、`Number.isFinite` 兜底，向后兼容旧 wrapper。
- 契约清晰：SQLite 仅补 usage、绝不改动 Step 边界；`turnAnchor` 5s 阈值防止错配上一 turn 旧 usage。
- 新增 retry/cold-start 两个测试套件，直接覆盖两条历史评论场景。

---
*评审报告详见: `code-review/pr-212/final-report.md`*
*Generated by LoongSuite-Pilot Code Review Agent*

File: assets/hooks/qoder-hook-processor.mjs (hunk 1 above)
Comment (linrunqi08) [reply assets/hooks/qoder-hook-processor.mjs]: **[fixed]** `buildLlmBoundaries` 已增加兜底 flush，未闭合的 group 不再永久悬挂。

**证据:** `completedByHook`（assets/hooks/qoder-hook-processor.mjs:775-783）在本 group 起始之后的 `PostToolUse`/`PostToolUseFailure` 完成信号数 ≥ `currentToolIds.size` 时强制 `flushGroup()`，覆盖「tool_use 被取消/中断、transcript 无对应 tool_result」；循环末尾无条件 `flushGroup()`（:794）兜底残留 group；回放用例 `tests/unit/hooks/qoder-hook-processor-retry.test.mjs:304`（uses PostToolUseFailure to close a tool cycle whose result is missing）直接覆盖。

---
*Generated by LoongSuite-Pilot Code Review Agent*

File: assets/hooks/qoder-hook-processor.mjs (hunk 2 above)
Comment (linrunqi08) [reply assets/hooks/qoder-hook-processor.mjs]: **[partially-fixed]** 核心多重读放大已消除，残留 size cap 未落实。

**证据:** 重试轮询（assets/hooks/qoder-hook-processor.mjs:230-248）每轮只 `readTranscriptSnapshot` 一次，同一 snapshot 复用于 `getLineRangeInfo`（第 4 参复用 lineCount）、`findTriggeredTurnWindow` 与 `processTranscript`（经 `opts.snapshot` 透传 :291），消除「window/lineCount/process 各全量读一次」的多重读放大；`findTriggeredTurnWindow` 从 `triggerEndLine` 向前回扫定位 prompt/Stop，非从 0 盲扫（建议 1、2 已落实）。未落实：建议 (3) 单次解析行数/字节上限——`readTranscriptSnapshot`（:574-594）仍全量 `readFileSync`+`split`+逐行 `JSON.parse`。该全量读属 base 既有特性、本 PR 读次数不增反减，残留仅 Low 级 hardening，不阻断合入，保持 open 跟踪 size cap 补强。

---
*Generated by LoongSuite-Pilot Code Review Agent*

**Review by linrunqi08 — APPROVED** [review]
Code Review Agent 裁决：**APPROVE-READY** ✅

无 Critical/High/Medium 阻断项，Lifecycle 三项全 PASS。历史评论 #3689728074 已修复并 resolve，#3689729644 核心 O(n²) 读放大已消除（残留 size-cap 为 Low，保持 open 跟踪）。残余风险均为 Low，不阻断合入。详见摘要评论与 `code-review/pr-212/final-report.md`。

---
*Generated by LoongSuite-Pilot Code Review Agent*

**Review by ralf0131 — APPROVED** [review]
### Summary
Stabilizes Qoder trace step assembly by fixing timestamp precision issues, transcript read patterns, and turn boundary detection. The core change is a single-read snapshot model that replaces multiple re-reads of the transcript file.

### Highlights
- **Sub-millisecond ordering**: `timestampOrderNanos()` preserves microsecond precision from Qoder timestamps, preventing adjacent LLM boundary collapse
- **Single-read snapshot**: `readTranscriptSnapshot()` reads once and supplies all downstream consumers — no more race conditions from re-reading
- **Fixed turn window**: Retry with `--trigger-end-line` pins to the exact turn that triggered it, preventing expansion into the next queued prompt
- **Progress event dedup**: Resets `lastProgressHookEvent` on content rows, preventing false duplicate suppression across turn boundaries
- **Stop reprocessing**: Correctly reprocesses from line 0 when Stop is detected, ensuring complete ReAct chain visibility

Well-tested with comprehensive edge cases for retry, incomplete transcript, and boundary detection.

---
*Automated review by github-manager-bot*
